src/containers/Surah/index.js

```diff
@@ -40,57 +40,112 @@ import { setOption, toggleReadingMode } from '../../redux/modules/options';
+              //this.debug('load surahs failed, trying again');
+              return dispatch(loadAll()).
+              catch((err) => {
+                //this.debug('throwing an err, what happens?');
+                throw `couldnt load surahs ${err}`
+              });
+            })/*.then(() => {this.debug('surahs loaded');})*/;
+        })
+        .then(() => {
+          if (range) {
+            if (range.includes('-')) {
+              [from, to] = range.split('-');
+            } else { // Single ayah. For example /2/30
+              from = range;
+              to = parseInt(range, 10) + ayahRangeSize;
+            }
 
-      if (!isLoaded(getState(), surahId, from, to)) {
-        dispatch(clearCurrent(surahId)); // In the case where you go to same surah but later ayahs.
+            if (isNaN(from) || isNaN(to)) {
+              // Something wrong happened like /2/SOMETHING
+              // going to rescue by giving beginning of surah.
+              [from, to] = [1, ayahRangeSize];
+            }
+          } else {
+            [from, to] = [1, ayahRangeSize];
+          }
 
-        return dispatch(loadAyahs(surahId, from, to, options));
-      }
+          try {
+            const state = getState();
+            from = Math.min(from, state.surahs.entities[surahId].ayat);
+              to = Math.min(to,   state.surahs.entities[surahId].ayat);
+          } catch(err) {
+            this.debug.err(`debug error ${err}`);
+          }
+        }, (err) => {
+          throw `could not load surahs`;
+        })
+        .then(() => {
+          if (!isLoaded(getState(), surahId, from, to)) { // In the case where you go to same surah but later ayahs.
+            return (new Promise((resolve, reject) => {
+              //this.debug('clearing surah ayahs and a brief 50ms pause before...');
+              dispatch(clearCurrent(surahId));
+              return setTimeout(() => {resolve()}, 50);
+            })).then(() => {
+              //this.debug('...loading ayahs');
+              return dispatch(loadAyahs(surahId, from, to, options)).
+              catch((err) => {
+                //this.debug('load ayahs failed, trying again');
+                return dispatch(loadAyahs(surahId, from, to, options)).
+                catch((err) => {
+                  throw `couldnt load ayahs ${err}`
+                });
+              }).then((args) => {
+                //this.debug(`ayahs loaded ${args}`);
+              });
+            }).catch((err) => {
+              throw `puking it up ${err}`;
+            });
+          }
+        })
+        .catch((err) => {
+          this.debug.err(`error was ${err}, going back home`);
+          return dispatch(push('/'))
+        });
+      };
+
+      this.debug('first promise start');
+      return p()
+        .then((args) => {
+          this.debug(`first promise end ${args}`);
+        })
+        .catch((err) => {
+          this.debug(`first promise err ${err}`);
+        });
```

Comment:
This whole method is a mess and really difficult to understand what is happening...